crates/astria-sequencer-relayer/src/relayer/celestia_client/builder.rs

```diff
@@ -0,0 +1,137 @@
+use std::sync::Arc;
+
+use astria_core::generated::{
+    celestia::v1::query_client::QueryClient as BlobQueryClient,
+    cosmos::{
+        auth::v1beta1::query_client::QueryClient as AuthQueryClient,
+        base::tendermint::v1beta1::{
+            service_client::ServiceClient as NodeInfoClient,
+            GetNodeInfoRequest,
+        },
+        tx::v1beta1::service_client::ServiceClient as TxClient,
+    },
+};
+use http::Uri;
+use tendermint::account::Id as AccountId;
+use tonic::transport::{
+    Channel,
+    Endpoint,
+};
+use tracing::trace;
+
+use super::{
+    super::State,
+    Bech32Address,
+    Bech32EncodeError,
+    CelestiaClient,
+    CelestiaKeys,
+    Error,
+    GrpcResponseError,
+};
+
+/// A builder for a [`CelestiaClient`].
+#[derive(Clone)]
+pub(in crate::relayer) struct Builder {
+    /// The inner `tonic` gRPC channel shared by the various generated gRPC clients.
+    grpc_channel: Channel,
+    /// The crypto keys associated with our Celestia account.
+    signing_keys: CelestiaKeys,
+    /// The Bech32-encoded address of our Celestia account.
+    address: Bech32Address,
+    /// A handle to the mutable state of the relayer.
+    state: Arc<State>,
+}
+
+impl Builder {
+    /// Returns a new `Builder`, or an error if Bech32-encoding the `signing_keys` address fails.
+    pub(in crate::relayer) fn new(
+        uri: Uri,
+        signing_keys: CelestiaKeys,
+        state: Arc<State>,
+    ) -> Result<Self, Error> {
+        let grpc_channel = Endpoint::from(uri).connect_lazy();
+        let address = bech32_encode(&signing_keys.address)?;
+        Ok(Self {
+            grpc_channel,
+            signing_keys,
+            address,
+            state,
+        })
+    }
+
+    /// Returns a new `CelestiaClient` initialized with info retrieved from the Celestia app.
+    ///
+    /// On failure, the builder is returned along with the error in order to support retrying.
+    pub(in crate::relayer) async fn try_build(self) -> Result<CelestiaClient, Error> {
+        let chain_id = self.fetch_chain_id().await?;
+
+        let Self {
+            grpc_channel,
+            signing_keys,
+            address,
+            state,
+        } = self;
+
+        let tx_client = TxClient::new(grpc_channel.clone());
+        let blob_query_client = BlobQueryClient::new(grpc_channel.clone());
+        let auth_query_client = AuthQueryClient::new(grpc_channel.clone());
+        let mut client = CelestiaClient {
+            grpc_channel,
+            tx_client,
+            blob_query_client,
+            auth_query_client,
+            signing_keys,
+            address,
+            chain_id,
+        };
+
+        client.fetch_and_cache_cost_params(state.clone()).await?;
+        state.set_celestia_connected(true);
```

Comment:
In the current code, this is only done after the first successful blob submission.  Here we're setting it before the first attempt, but I don't think it'll be an issue since we have at least had successful comms with the Celestia app by now (that happens inside `client.fetch_and_cache_cost_params` above).  Just wanted to call this out as it's a slight change from current behaviour.